### PR TITLE
IGNITE-17777 Thin 3.0: use BinaryTuple for Compute and SQL results and arguments

### PR DESCRIPTION
**Pull request by ptupitsyn**
Use BinaryTuple for:
* Compute arguments and results
* SQL arguments, results, and properties

As a result, MsgPack is no longer used to encode user data and does not need to support Ignite data types. Unused code removed.

.NET, Java, JDBC clients updated.

Additionally:
* Fixed decimal serialization in .NET

### CODE REVIEW

File: modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java [hunk 1]
```diff
@@ -49,72 +79,251 @@ public static void readAndSetColumnValue(
         }
 
         switch (clientDataType) {
-            case ClientDataType.INT8:
+            case INT8:
```
Comment (isapego) [inline modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java:82]: Is importing all the enum values separately required by the coding convention?

File: modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java [hunk 2]
```diff
@@ -49,72 +79,251 @@ public static void readAndSetColumnValue(
...
+     */
+    public static void appendObject(BinaryTupleBuilder builder, Object obj) {
+        if (obj == null) {
+            builder.appendNull();
+            builder.appendInt(0);
+            builder.appendNull();
+        } else if (obj instanceof Byte) {
+            builder.appendInt(ClientDataType.INT8);
+            builder.appendInt(0);
+            builder.appendByte((Byte) obj);
```
Comment (isapego) [inline modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java:243]: Are those `type`, `scale` and `value`? Maybe we should add a function with 3 args just so it's more clear what is written.

File: modules/platforms/dotnet/Apache.Ignite/Internal/Proto/BinaryTuple/BinaryTupleBuilder.cs [hunk 3]
```diff
@@ -534,6 +518,126 @@ public void AppendObject(object? value, ClientDataType colType, int scale)
...
+        /// </summary>
+        /// <param name="value">Value.</param>
+        public void AppendObjectWithType(object? value)
+        {
+            switch (value)
+            {
+                case null:
+                    AppendNull();
+                    AppendNull();
+                    AppendNull();
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite/Internal/Proto/BinaryTuple/BinaryTupleBuilder.cs:532]: The same comment as for Java. Also, there is some mismatch to Java packing for `null`

**Review by isapego — APPROVED** [review]

File: modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java:82]: Coding convention does not mandate this, as far as I understand.

File: modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java (hunk 2 above)
Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientBinaryTupleUtils.java]: Good point, done.

File: modules/platforms/dotnet/Apache.Ignite/Internal/Proto/BinaryTuple/BinaryTupleBuilder.cs (hunk 3 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/Internal/Proto/BinaryTuple/BinaryTupleBuilder.cs]: Fixed.

**Review by isapego — APPROVED** [review]

**Pull request merged**